File: app/src/main/java/com/example/visiba/visibarectest/Activites/MainActivity.java
package com.example.visiba.visibarectest.Activites;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ListView;
import android.widget.RelativeLayout;

import com.example.papersoccer.visibarectest.R;
import com.example.visiba.visibarectest.Activites.Abstractions.BaseActivity;
import com.example.visiba.visibarectest.Adapters.WallPostAdapter;
import com.example.visiba.visibarectest.AppImage;
import com.example.visiba.visibarectest.Enums.ImageRequestCodeEnum;
import com.example.visiba.visibarectest.Repositories.AppImageRepository;
import com.example.visiba.visibarectest.WallPost;
import com.example.visiba.visibarectest.Repositories.WallPostRepository;

import java.util.ArrayList;

public class MainActivity extends BaseActivity {

    EditText newPostInput;
    RelativeLayout newPostButtonsLayout;
    ListView wallPostsListView;

    WallPostRepository wallPostRepository;
    AppImageRepository appImageRepository;

    ImageButton leftImageButton;
    ImageButton rightImageButton;

    AppImage leftImage;
    AppImage rightImage;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        newPostInput = (EditText)findViewById(R.id.newPostInput);
        wallPostsListView = (ListView)findViewById(R.id.wallPostsListView);
        wallPostsListView.addHeaderView(new View(this));

        leftImageButton = (ImageButton)findViewById(R.id.leftImageButton);
        rightImageButton = (ImageButton)findViewById(R.id.rightImageButton);

        newPostButtonsLayout = (RelativeLayout)findViewById(R.id.newPostButtonsLayout);

        appImageRepository = new AppImageRepository(this);
        wallPostRepository = new WallPostRepository(this, appImageRepository);

        populateWallPostsListView();

        newPostInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void afterTextChanged(Editable s) {

            }

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if(s.length() != 0) newPostButtonsLayout.setVisibility(View.VISIBLE);
                else {
                    newPostButtonsLayout.setVisibility(View.INVISIBLE);
                }
            }
        });
    }

    public void onSendWallPostClick(View view)
    {
        WallPost wallPost = new WallPost(newPostInput.getText().toString(), leftImage, rightImage);
        wallPostRepository.SaveWallPostData(new WallPost.SerializableWallPost(wallPost));
        recreate();
    }

    public void OpenCameraActivity(int imageButtonRequestCode)
    {
        Intent intent = new Intent(this, CameraActivity.class);
        intent.putExtra("IMAGE_BUTTON_REQUEST_CODE", imageButtonRequestCode);
        startActivityForResult(intent, imageButtonRequestCode);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == ImageRequestCodeEnum.LEFT) {
            if (resultCode == RESULT_OK) {
                String imageId = data.getStringExtra("IMAGE_ID");
                leftImage = setImageButtonPreviewImage(leftImageButton, imageId);;
            }
        }
        else if (requestCode == ImageRequestCodeEnum.RIGHT) {
            if (resultCode == RESULT_OK) {
                String imageId = data.getStringExtra("IMAGE_ID");
                rightImage = setImageButtonPreviewImage(rightImageButton, imageId);
            }
        }
    }

    private AppImage setImageButtonPreviewImage(ImageButton imageButton, String imageId)
    {
        AppImage appImage = appImageRepository.loadImagesFromStorage(imageId);
        imageButton.setImageDrawable(appImage.drawableImage);
        imageButton.setImageTintList(null); // White Tint
        return appImage;
    }

    public void onLeftImageButtonClick(View v)
    {
        try {
            OpenCameraActivity(ImageRequestCodeEnum.LEFT);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void onRightImageButtonClick(View v)
    {
        try {
            OpenCameraActivity(ImageRequestCodeEnum.RIGHT);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void populateWallPostsListView()
    {
        ArrayList<WallPost> wallPosts = new ArrayList<>();

        for (WallPost.SerializableWallPost serializableWallPost : wallPostRepository.LoadAllWallPostsFromStorage())
        {
            WallPost wallPost = wallPostRepository.ConvertSerializableToWallPost(serializableWallPost);
            wallPosts.add(wallPost);
        }

        WallPostAdapter wallPostAdapter = new WallPostAdapter(getBaseContext(), wallPosts);
        wallPostsListView.setAdapter(wallPostAdapter);
    }
}

File: app/src/main/java/com/example/visiba/visibarectest/Activities/MainActivity.kt
package com.example.visiba.visibarectest.Activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.widget.ImageButton
import com.example.papersoccer.visibarectest.R
import com.example.visiba.visibarectest.Activities.Abstractions.BaseActivity
import com.example.visiba.visibarectest.Adapters.WallPostAdapter
import com.example.visiba.visibarectest.AppImage
import com.example.visiba.visibarectest.Enums.ImageRequestCodeEnum
import com.example.visiba.visibarectest.Repositories.AppImageRepository
import com.example.visiba.visibarectest.WallPost
import com.example.visiba.visibarectest.Repositories.WallPostRepository
import io.paperdb.Paper
import kotlinx.android.synthetic.main.content_main.*

class MainActivity : BaseActivity() {

    val newPostButtonsLayout get() = main_newPostButtonsLayout
    val wallPostsListView get() = main_wallPostsListView

    val newPostInput get() = main_newPostInput

    lateinit var wallPostRepository: WallPostRepository
    lateinit var appImageRepository: AppImageRepository

    val leftImageButton get() = main_leftImageButton
    val rightImageButton get() = main_rightImageButton

    var leftImage: AppImage? = null
    var rightImage: AppImage? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        init()
    }

    private fun init() {
        setContentView(R.layout.activity_main)
        val toolbar = findViewById(R.id.toolbar) as Toolbar
        setSupportActionBar(toolbar)

        Paper.init(this)

        appImageRepository = AppImageRepository()
        wallPostRepository = WallPostRepository()

        populateWallPostsListView()

        newPostInput.addTextChangedListener(object : TextWatcher {
            override fun afterTextChanged(s: Editable) {

            }
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (s.length != 0)
                    newPostButtonsLayout.visibility = View.VISIBLE
                else {
                    newPostButtonsLayout.visibility = View.INVISIBLE
                }
            }
        })
    }

    fun onSendWallPostClick(view: View) {
        val wallPost = WallPost(newPostInput.text.toString(), leftImage!!, rightImage!!)
        wallPostRepository.SaveWallPostData(wallPost)
        init()
    }

    fun OpenCameraActivity(imageButtonRequestCode: Int) {
        val intent = Intent(this, CameraActivity::class.java)
        intent.putExtra("IMAGE_BUTTON_REQUEST_CODE", imageButtonRequestCode)
        startActivityForResult(intent, imageButtonRequestCode)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        if (requestCode == ImageRequestCodeEnum.LEFT) {
            if (resultCode == Activity.RESULT_OK) {
                val imageId = data.getStringExtra("IMAGE_ID")
                leftImage = setImageButtonPreviewImage(leftImageButton, imageId)!!
            }
        } else if (requestCode == ImageRequestCodeEnum.RIGHT) {
            if (resultCode == Activity.RESULT_OK) {
                val imageId = data.getStringExtra("IMAGE_ID")
                rightImage = setImageButtonPreviewImage(rightImageButton, imageId)!!
            }
        }
    }

    private fun setImageButtonPreviewImage(imageButton: ImageButton, imageId: String): AppImage? {
        val appImage = appImageRepository.loadItem(imageId)
        imageButton.setImageDrawable(appImage?.getDrawable(this))
        imageButton.imageTintList = null // White Tint
        return appImage
    }

    fun onLeftImageButtonClick(v: View) {
        try {
            OpenCameraActivity(ImageRequestCodeEnum.LEFT)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun onRightImageButtonClick(v: View) {
        try {
            OpenCameraActivity(ImageRequestCodeEnum.RIGHT)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun populateWallPostsListView() {
        val wallPosts = wallPostRepository.loadAllItems()
        val wallPostAdapter = WallPostAdapter(baseContext, wallPosts!!.asReversed())
        wallPostsListView.adapter = wallPostAdapter
    }
}
